core/src/main/java/org/infinispan/atomic/impl/AtomicKeySetImpl.java

```diff
@@ -0,0 +1,557 @@
+      }
+
+      @Override
+      public void writeTo(ObjectOutput output) throws IOException {
+         output.writeObject(group);
+         output.writeObject(key);
+      }
+
+      @Override
+      public String toString() {
+         return "AtomicKeySetImpl.Key{group=" + group + ", key=" + key + "}";
+      }
+   }
+
+   static class ReadAll<K> implements Function<EntryView.ReadEntryView<Object, Object>, Set<K>>, Externalizable {
+      private static final ReadAll INSTANCE = new ReadAll();
+
+      static <K> ReadAll<K> instance() {
+         return INSTANCE;
+      }
+
+      @Override
+      public Set<K> apply(EntryView.ReadEntryView<Object, Object> view) {
+         return view.find().map(value -> {
+            if (value instanceof AtomicKeySetImpl) {
+               return ((AtomicKeySetImpl) value).toSet();
+            } else {
+               throw log.atomicMapHasWrongType(value, AtomicKeySetImpl.class);
+            }
+         }).orElse(null);
+      }
+
+      @Override
+      public Type type() {
+         return Type.READ_ALL;
+      }
+
+      @Override
+      public void writeTo(ObjectOutput output) throws IOException {
+      }
+   }
+
+   static class Touch<MK> implements Function<EntryView.ReadWriteEntryView<MK, Object>, Void>, Externalizable {
+      private final ByteString cacheName;
+
+      public Touch(ByteString cacheName) {
+         this.cacheName = cacheName;
+      }
+
+      public static <MK> Touch readFrom(ObjectInput input) throws IOException, ClassNotFoundException {
+         ByteString cacheName = ByteString.readObject(input);
+         return new Touch(cacheName);
+      }
+
+      @Override
+      public Void apply(EntryView.ReadWriteEntryView<MK, Object> view) {
+         if (view.find().isPresent()) {
+            Object value = view.find().get();
+            // Trying to touch when the context contains a delta shouldn't pose a problem
+            if (!(value instanceof AtomicKeySetImpl)) {
+               throw log.atomicMapHasWrongType(value, AtomicKeySetImpl.class);
+            }
+         } else {
+            view.set(new AtomicKeySetImpl(cacheName, view.key(), Collections.emptySet(), null, null));
+         }
+         return null;
+      }
+
+      @Override
+      public Type type() {
+         return Type.TOUCH;
+      }
+
+      @Override
+      public void writeTo(ObjectOutput output) throws IOException {
+         ByteString.writeObject(output, cacheName);
+      }
+   }
+
+   static class Add<K> implements Function<EntryView.ReadWriteEntryView<Object, Object>, Void>, Externalizable {
+      private final K key;
+
+      public Add(K key) {
+         this.key = key;
+      }
+
+      public static <K> Add readFrom(ObjectInput input) throws IOException, ClassNotFoundException {
+         return new Add((K) input.readObject());
+      }
+
+      @Override
+      public Void apply(EntryView.ReadWriteEntryView<Object, Object> view) {
+         if (view.find().isPresent()) {
```

Comment:
Since you call `view.set()`, that would create another object capturing the `view`.